executor/builder.go

```diff
@@ -3462,25 +3473,51 @@ func (b *executorBuilder) buildMPPGather(v *plannercore.PhysicalTableReader) Exe
 		virtualColumnRetFieldTypes: []*types.FieldType{},
 	}
 
+	gather.memTracker.AttachTo(b.ctx.GetSessionVars().StmtCtx.MemTracker)
+
 	var hasVirtualCol bool
 	for _, col := range v.Schema().Columns {
 		if col.VirtualExpr != nil {
 			hasVirtualCol = true
 			break
 		}
 	}
-	if hasVirtualCol {
-		// If hasVirtualCol, Join should not pushdown to tiflash,
-		// so there is only one TableScan.
-		ts, err := v.GetTableScan()
-		if err != nil {
-			b.err = err
+
+	var isSingleDataSource bool
+	tableScans := v.GetTableScans()
+	if len(tableScans) == 1 {
+		isSingleDataSource = true
+	}
+
+	// 1. hasVirtualCol: when got virtual column in TableScan, will generate plan like the following,
+	//                   and there will be no other operators in the MPP fragment.
+	//     MPPGather
+	//       ExchangeSender
+	//         PhysicalTableScan
+	// 2. UnionScan: there won't be any operators like Join between UnionScan and TableScan.
+	//               and UnionScan cannot push down to tiflash.
+	if !isSingleDataSource {
+		if hasVirtualCol || b.encounterUnionScan {
+			b.err = errors.Errorf("should only have one TableScan in MPP fragment(hasVirtualCol: %v, encounterUnionScan: %v)", hasVirtualCol, b.encounterUnionScan)
 			return nil
 		}
-		gather.columns = ts.Columns
+		return gather
+	}
+
+	// Setup MPPGather.table if isSingleDataSource.
+	// Virtual Column or UnionScan need to use it.
+	ts := tableScans[0]
+	gather.columns = ts.Columns
+	if hasVirtualCol {
 		gather.virtualColumnIndex, gather.virtualColumnRetFieldTypes = buildVirtualColumnInfo(gather.Schema(), gather.columns)
 	}
-	gather.memTracker.AttachTo(b.ctx.GetSessionVars().StmtCtx.MemTracker)
+	tbl, _ := b.is.TableByID(ts.Table.ID)
+	isPartition, physicalTableID := ts.IsPartition()
+	if isPartition {
```

Comment:
These code should only work for static partition prune? Maybe add a comment here.

Reply:
I need take a look.